### ISSUE TITLE
Provide data/event flow examples and facilities

### ISSUE THREAD

**Issue by JoshMarler** [nick-thompson/blueprint#66]
Currently `ReactApplicationRoot` and `EcmascriptEngine` support pushing events from native/C++ code into JS/React via the `dispatchEvent` mechanism. 

`dispatchEvent` is a function registered/created against the global `__BlueprintNative__` object by EventBridge.js. 

It seems that a lot of apps and client code would benefit from a generic "event based" framework or pattern to communicate between C++ and JS. 

For example, it would be nice to both send an `event` (consisting of an ID and payload/args) from C++ to Javascript and from Javascript to C++. 

I might wish subscribe to a "parameterChanged" event in JS and also trigger/signal that same event back to C++.

Is this something that blueprint could/should offer via `EcmascriptEngine`/`ReactApplicationRoot`?

I could imagine a complimentary function to `ReactApplicationRoot::dispatchEvent` along the lines of `ReactApplicationRoot::handleEvent`.

i.e.

```
virtual void ReactApplicationRoot::handleEvent(const juce::String& eventType, const juce::NativeFunctionArgs& args) = 0;
```

This might require some change or "fleshing out" of an `EventBridge`  to support a `publish` and `subscribe` like interface. 

@nick-thompson I'd love to hear your ideas around this area. I'm going to imnplement some sort of event based communication in my app over the coming days so will let you know if I come up with any interesting stuff.

**Comment by JoshMarler** [comment]
@nick-thompson, in the middle of implementing two-way event dispatch at the moment as a test so I can open up a PR if you like the sound of this idea?

**Comment by JoshMarler** [comment]
@nick-thompson, FYI I've got this working locally and so far seems pretty  good.

**Comment by nick-thompson** [comment]
Hmmm a lot to think about here, thanks for starting the conversation @JoshMarler.

> The big initial question feels like "Should Blueprint provide this or should examples be given with data-flow patterns/implementations left to the users of Blueprint?"

I think that's exactly the right place to start. I suppose my initial stance is that Blueprint itself shouldn't take a stance on data flow patterns, because every app is different and everbody has their own opinions on data flow patterns. In React you see a ton of Flux/Redux, but there are plenty of people who prefer an RxJS style approach, for example. However, I think Blueprint _should_ provide _some_ set of facilities for easily implementing your data flow pattern of choice.

`EventBridge` was an early idea and very much only a partially formed idea. It came about as I was building a Flux pattern for Remnant in Blueprint, where my one true data store was the AudioProcessorValueTreeState, all changes were broadcast as events into js land, and from js there were no direct ways to manipulate state, only to issue "actions," as in the Flux model, such as "parameter value update." That's why you only see the C++->JS direction. That's nice for Flux and for audio plugins, but I can imagine many apps even with a similar data flow pattern where you'd want the "one source of truth" living in js land and broadcasting change events to C++. So, thinking about that more now, I think your idea for building out `EventBridge` into a true pub/sub system that can go either direction is the right move.

I think that also provides a good opportunity to implement something I've done a few times but not quite yet integrated fully: a coherent way to throttle events going from C++->JS. Right now the GainPlugin calls into js on every parameter value change, and this event comes in _all the time_ during a drag gesture or during host automation. I implemented what I called a "throttle map" (https://github.com/nick-thompson/blueprint/blob/master/blueprint/core/blueprint_ThrottleMap.h) that I think we could integrate into the EventBridge from C++->JS so that as you dispatch events you can choose if they should be throttled or not. Need to spend a minute on what that API might look like from the ReactAppRoot side of things.

Anyway, then carrying on to the automatic ValueTree 2-way sync– I think that's an awesome idea, but my initial reaction on it is that that falls into the "taking a stance on data flow patterns" category (at least, 2-way data binding is often quite at odds with the Flux model). Which makes me think that maybe it shouldn't be part of blueprint core (but absolutely should be a well documented example at least). I need to spend a little more time on it, and if you have some code ready I'd love to see it, that'd be quite helpful. Either way, I think you're totally right that many people would find that super helpful, and it would apply to many apps.

Hope that was a sensible ramble. I'll think on it some more, revisit tomorrow and look more closely at the API design side of this

**Comment by JoshMarler** [comment]
Hey Nick,

Thanks for the discussion. Yep fair enough, I think I agree. I'd be really pleased just to see the two way pub sub mechanism there. I have that working at the moment but it's just a bare bones implementation for me to leverage. I'll think about the API some more today and maybe do the usual "open up a PR for Nick to fix :-)"

 I've been down the redux route a couple of times in the past and found it "a bit wordy". I've more experience with Angular and at the time (when I was actually writing some JavaScript) it had a fairly opinionated rxjs approach which I quite liked.

I'm still very much in the exploratory phased with this stuff so I'll probably try a few different patterns. Flux is a thing I should probably catch back up on. 

A value tree like system appeals as I'm working with a pretty big application with LOTS of events. Being able to bind events straight into some sort of "model" interface where each event ID maps to a model/ValueTree property key could save us a lot of repetitive boiler plate.

However, totally possible this approach is going to be too fixed in some cases and result in too much runtime type checking. 

I suspect I'll also need multiple options so I agree, Blueprint probably shouldn't prescribe any ValueTree based approach, it should.just provide thing like the EventBridge to facilitate any pattern users might take. ValueTree binding will probably just make a nice example in the docs.

**Comment by nick-thompson** [comment]
> I'd be really pleased just to see the two way pub sub mechanism there. I have that working at the moment but it's just a bare bones implementation for me to leverage. I'll think about the API some more today and maybe do the usual "open up a PR for Nick to fix :-)"

Yea same, I think the two-way pub sub is definitely the right move. Please do share some code when you have it! I'll spend some time thinking on the API myself as well. I feel like it could really be quite simple

> I've been down the redux route a couple of times in the past and found it "a bit wordy".

Definitely. I'm personally not a huge fan of redux, but so many people are. I often write an as-minimal-as-possible Flux pattern (as in the GainPlugin example) for my state management for smaller apps.

> A value tree like system appeals as I'm working with a pretty big application with LOTS of events. Being able to bind events straight into some sort of "model" interface where each event ID maps to a model/ValueTree property key could save us a lot of repetitive boiler plate.

Right, yea, I can imagine many cases where this auto-enumerating event binding makes tons of sense. And this morning as I'm thinking through it more, while I do still think that Blueprint should probably only provide the EventBridge to facilitate this, this kind of value tree event dispatcher/synchronizer might make a great add to the "Extras" library in #53. Especially if our extras library will include bindings to juce-specific things like the juce::Slider, bindings to a juce::ValueTree seems to make a lot of sense there. I'm picturing something like:

```
class MainComponent   : public Component
{
public:
    //==============================================================================
    MainComponent()
        : treeSync(appRoot, state) // Automatically sets up listeners to call `appRoot.dispatchEvent` and receive events from js
    {
        addAndMakeVisible(appRoot);
        setSize(400, 300);
    }

    ~MainComponent();

    //==============================================================================
    void paint (Graphics&) override;

    void resized() override
    {
        appRoot.setBounds(getLocalBounds());
    }

private:
    //==============================================================================
    juce::ValueTree state;
    blueprint::ReactApplicationRoot appRoot;
    blueprint::extras::ValueTreeSynchronizer treeSync;

    JUCE_DECLARE_NON_COPYABLE_WITH_LEAK_DETECTOR (MainComponent)
};
```

Curious if that's along the lines of what you've already written? I'm obviously hand-waving but the thought is that, since this lives on top of the EventBridge layer, it could very much be something that people take off the shelf if they want it, or leave it if they want some specific Redux/Flux/RXJS/Etc

**Comment by JoshMarler** [comment]
@nick-thompson, Yes! Totally. The extras module sounds like a great place for a ValueTreeSynchronizer. I'm literally working on things like this right now. 

I've got a very simple two way EventBridge in js which is working just fine but needs a littler more thought. I'm experimenting with the dispatchEvent interface at the moment so I'll have something to send over soon. I'm also building some JS components which can hook into this whole system so the "extras" module feels like something that could definitely start taking shape soon. 

Two-way EventBridge:

```
import EventEmitter from 'events';

class EventBridge extends EventEmitter {
  constructor() {
    super();

    this.emit = this.emit.bind(this);

    this.emittingFromNative = false;

    // TODO: Make constructor arg/configurable?
    this.setMaxListeners(30);

    __BlueprintNative__.dispatchEventFromNative = (eventType, ...args) => {
      this.emittingFromNative = true;
      this.emit(eventType, ...args);
      this.emittingFromNative = false;
    }
  }

  // Override EventEmitter.emit()
  emit(eventType, ...args) {
    if (this.emittingFromNative) {
      super.emit(eventType, ...args);
    }
    else {
      __BlueprintNative__.dispatchEventToNative(eventType, ...args);
    }
  }
}

const _singletonInstance = new EventBridge();
export default _singletonInstance;
```

Can totally send over a PR soon with the pub/sub stuff. ValueTree utility class might take me a little longer :-) We aren't actually using ValueTree at the moment so I've bound events to our own "data-model" type of class but the pattern follows the exact same principles that binding to a ValueTree would. I'm experimenting with a couple of different data-model interfaces and ValueTree is the next on my list to tackle so if I get something nice together I shall send it over to you. 

Cheers.
